go_stash/groups_test.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2021 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gostash
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+)
+
+func TestGetGroup(t *testing.T) {
+	tests := []struct {
+		name      string
+		groupName string
+		output    string
+	}{
+		{
+			name:      "test group does not exist",
+			groupName: "admin",
+		},
+		{
+			name:      "test a group",
+			groupName: "dev-1",
+		},
+	}
+
+	validNames := []string{"dev-1"}
+
+	mux, server, client := setup(t)
+	defer teardown(server)
+
+	path := fmt.Sprintf("%s/%s", stashURIprefix, groupsURI)
+	mux.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
+		for _, substr := range validNames {
+			if strings.Contains(r.Form.Get("filter"), substr) {
+				w.WriteHeader(http.StatusOK)
+				u := &Group{
+					Name: substr,
+				}
+				json.NewEncoder(w).Encode(u)
+				return
+			}
+		}
+
+		http.Error(w, "The specified group does not exist", http.StatusNotFound)
+
+		return
+
+	})
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			group, err := client.Groups.GetGroup(ctx, tt.groupName)
+			if err != nil {
+				if err != ErrNotFound {
+					t.Fatalf("Groups.GetGroup returned error: %v", err)
+				}
+			} else {
+				if group.Name != tt.groupName {
+					t.Errorf("Groups.GetGroup returned group %s, want %s", group.Name, tt.groupName)
+				}
+			}
+
+		})
+	}
+}
+
+func TestListGroup(t *testing.T) {
+	gNames := []string{"avengers", "x-men", "ff4", "x-team"}
+
+	mux, server, client := setup(t)
+	defer teardown(server)
+
+	path := fmt.Sprintf("%s/%s", stashURIprefix, groupsURI)
+	mux.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		u := struct {
+			Groups []Group `json:"values"`
+		}{[]Group{
+			{
+				Name: gNames[0],
+			},
+			{
+				Name: gNames[1],
+			},
+			{
+				Name: gNames[2],
+			},
+			{
+				Name: gNames[3],
+			},
+		}}
+		json.NewEncoder(w).Encode(u)
+		return
+
+	})
+	ctx := context.Background()
+	list, err := client.Groups.List(ctx, nil)
+	if err != nil {
+		t.Fatalf("Groups.GetGroupList returned error: %v", err)
+	} else {
```

Comment:
You don't need this `else` because the test will already have stopped at the `t.Fatalf` above.